### PR TITLE
ARROW-13111: [R] altrep vectors for ChunkedArray

### PR DESCRIPTION
**Pull request by romainfrancois**
``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

(c_str <- ChunkedArray$create(c("un", "deux"), c("trois", "quatre")))
#> ChunkedArray
#> [
#>   [
#>     "un",
#>     "deux"
#>   ],
#>   [
#>     "trois",
#>     "quatre"
#>   ]
#> ]
(v_str <- c_str$as_vector())
#> [1] "un"     "deux"   "trois"  "quatre"

.Internal(inspect(v_str))
#> @7ff3ff98f540 16 STRSXP g0c0 [REF(65535)] arrow::ChunkedArray<0x7ff3f64a2128, string, 2 chunks, 0 nulls> len=4
```

<sup>Created on 2021-10-08 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

Only dealing with strings for now, int and double coming next.

### CODE REVIEW

**Comment by nealrichardson** [comment]
@romainfrancois maybe I misread, but it looks like everywhere that altrep was previously working with Arrays, it now does ChunkedArrays instead? So somewhere I guess (I didn't follow exactly where) when I do `array$as_vector()`, that makes a ChunkedArray with 1 chunk and sticks it in the altrep object?

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-versions

**Comment by nealrichardson** [comment]
> The downside here is that `Array$create(x)` might have to deal with the situation of `x` being an altrep backed by a `ChunkedArray` of n chunks, but it wants to create an `Array` so in that case a concatenation is done (this should not happen much).

Right, and in the cases where this might happen (e.g. you've done `read_parquet()` and get a data.frame of altrep vectors backed by ChunkedArrays, and then somehow you choose to turn that data.frame into a RecordBatch), you otherwise would have done that concatenation already when you made the data.frame. So this doesn't concern me.

> Perhaps we could have some R visible function that would either make an `Array` or a `ChunkedArray`.

I'm not sure how useful that would be. There was/is a part of the pyarrow converter that could return either Array or ChunkedArray (I think this was with strings, and instead of promoting to LargeString they were chunking), and I believe this caused lots of headaches.

**Comment by jonkeane** [comment]
Thinking about this a bit more and that it's on the file-write benchmark, this might be an issue with altrep -> R -> altrep (or sans the -> R in the middle) issue. We could try [turning altrep off in arrowbench's setup step](https://github.com/ursacomputing/arrowbench/blob/main/R/bm-write-file.R#L18). (I can do this, though I'm not sure when I'll have a chance to)

**Comment by jonkeane** [comment]
OK, I ran the arrowbench benchmarks locally, and indeed, the first from-data.frame writing one was what hung:

```
Running source=fanniemae_2016Q4 format=parquet compression=uncompressed input=data_frame lib_path=remote-romainfrancois/arrow@ARROW_13111_chunked_array_altrep cpu_count=8 name=write_file
```

I suspect this has to do with the fact that these data.frames are now back by altrep, and something funny is going on when we write them. I will try and reproduce them tomorrow, but I would expect `df <- read_parquet(...); write_parquet(...)` to hang. I'll make a PR to not use altrep when reading in data in arrowbench, but I suspect we want to fix this hang regardless.

**Comment by romainfrancois** [comment]
This is this very recent commit in R: https://github.com/r-devel/r-svn/commit/1c6226a0cf7375511ae1ce5708d711158a38e2eb, I'll move this out of snapshot

File: r/src/altrep.cpp [hunk 1]
```diff
@@ -215,9 +219,10 @@ struct AltrepVectorPrimitive : public AltrepVectorBase<AltrepVectorPrimitive<sex
 
   // The value at position i
   static c_type Elt(SEXP alt, R_xlen_t i) {
-    const auto& array = GetArray(alt);
-    return array->IsNull(i) ? cpp11::na<c_type>()
-                            : array->data()->template GetValues<c_type>(1)[i];
+    auto array = GetChunkedArray(alt)->Slice(i, 1)->chunk(0);
```
Comment (pitrou) [inline r/src/altrep.cpp:222]: Slicing is quite expensive for this. You should instead resolve the chunk yourself.

File: r/src/altrep.cpp [hunk 2]
```diff
@@ -237,29 +242,35 @@ struct AltrepVectorPrimitive : public AltrepVectorBase<AltrepVectorPrimitive<sex
...
     // This only materialize the region, into buf. Not the entire vector.
-    auto slice = GetArray(alt)->Slice(i, n);
+    auto slice = GetChunkedArray(alt)->Slice(i, n);
     R_xlen_t ncopy = slice->length();
 
-    // first copy the data buffer
-    memcpy(buf, slice->data()->template GetValues<c_type>(1), ncopy * sizeof(c_type));
+    c_type* out = buf;
+    for (const auto& array : slice->chunks()) {
+      auto n = array->length();
```
Comment (pitrou) [inline r/src/altrep.cpp:250]: Call this something else? It conflicts with the `n` function argument, which is error-prone.

File: r/tests/testthat/test-altrep.R [hunk 3]
```diff
@@ -91,11 +91,37 @@ test_that("empty vectors are not altrep", {
...
+  expect_true(is_arrow_altrep(cv))
+
+  # chunked array with 2 chunks
+  c_int <- ChunkedArray$create(0L, c(1L, NA, 3L))
+  c_dbl <- ChunkedArray$create(0, c(1, NA, 3))
+  expect_equal(c_int$num_chunks, 2L)
+  expect_equal(c_dbl$num_chunks, 2L)
+
+  expect_true(is_arrow_altrep(as.vector(c_int)))
+  expect_true(is_arrow_altrep(as.vector(c_dbl)))
```
Comment (pitrou) [inline r/tests/testthat/test-altrep.R:121]: Should you also test with a 0-chunk chunked array?

File: r/src/altrep.cpp (hunk 1 above)
Comment (romainfrancois) [reply r/src/altrep.cpp:222]: Thanks. I added this helper struct: 

```cpp
struct ArrayResolve {
  ArrayResolve(const std::shared_ptr<ChunkedArray>& chunked_array, int64_t i) {
    for (int idx_chunk = 0; idx_chunk < chunked_array->num_chunks(); idx_chunk++) {
      std::shared_ptr<Array> chunk = chunked_array->chunk(idx_chunk);
      auto chunk_size = chunk->length();
      if (i < chunk_size) {
        index_ = i;
        array_ = chunk;
        break;
      }

      i -= chunk_size;
    }
  }

  std::shared_ptr<Array> array_;
  int64_t index_;
};
```

File: r/tests/testthat/test-altrep.R (hunk 3 above)
Comment (romainfrancois) [reply r/tests/testthat/test-altrep.R:131]: ✔️

**Comment by romainfrancois** [comment]
@nealrichardson I believe this deals with https://issues.apache.org/jira/browse/ARROW-14706?filter=12346440, i.e. `test_same_Array()` is gone, and `$Same()` is. 

``` r
library(arrow, warn.conflicts = FALSE)
#> See arrow_info() for available features

a <- ChunkedArray$create(c(1, 2, 3), c(4, 5, 6))
b <- ChunkedArray$create(c(7, 8, 9))
c <- Array$create(c(10, 11, 12))
d <- Array$create(c(13, 14, 15))
e <- ChunkedArray$create(c(16, 17), c(18, 19))

x <- ChunkedArray$create(
  # converter to R vectors (with altrep but keeping underlying arrays)
  a$as_vector(), # 2 chunks
  b$as_vector(), # 1 chunk
  c$as_vector(), # 1 array
  
  # passed in directly
  d,
  e
)

x$chunk(0)$Same(a$chunk(0))
#> [1] TRUE
x$chunk(1)$Same(a$chunk(1))
#> [1] TRUE
x$chunk(2)$Same(b$chunk(0))
#> [1] TRUE
x$chunk(3)$Same(c)
#> [1] TRUE
x$chunk(4)$Same(d)
#> [1] TRUE
x$chunk(5)$Same(e$chunk(0))
#> [1] TRUE
x$chunk(6)$Same(e$chunk(1))
#> [1] TRUE
```

<sup>Created on 2021-11-18 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1.9000)</sup>

**Review by pitrou — APPROVED** [review]

**Comment by pitrou** [comment]
@romainfrancois @nealrichardson I'll let you close  https://issues.apache.org/jira/browse/ARROW-14706 if necessary.

**Pull request closed without merging**